openstack/orchestration/v1/resourcetypes/requests.go

```diff
@@ -0,0 +1,120 @@
+type SupportStatus string
+
+var (
+	// SupportStatusUnknown is returned when the resource type does not have a
+	// support status.
+	SupportStatusUnknown SupportStatus = "UNKNOWN"
+	// SupportStatusSupported indicates a resource type that is expected to
+	// work.
+	SupportStatusSupported SupportStatus = "SUPPORTED"
+	// SupportStatusDeprecated indicates a resource type that is in the process
+	// being removed, and may or may not be replaced by something else.
+	SupportStatusDeprecated SupportStatus = "DEPRECATED"
+	// SupportStatusHidden indicates a resource type that has been removed.
+	// Existing stacks that contain resources of this type can still be
+	// deleted or updated to remove the resources, but they may not actually
+	// do anything any more.
+	SupportStatusHidden SupportStatus = "HIDDEN"
+	// SupportStatusUnsupported indicates a resource type that is provided for
+	// preview or other purposes and should not be relied upon.
+	SupportStatusUnsupported SupportStatus = "UNSUPPORTED"
+)
+
+// ListOptsBuilder allows extensions to add additional parameters to the
+// List request.
+type ListOptsBuilder interface {
+	ToResourceTypeListQuery() (string, error)
+	ReturnDescription() bool
+}
+
+// ListOpts allows the filtering of collections through the API.
+type ListOpts struct {
+	// Filters the resource type list by a regex on the name.
+	Pattern string `q:"name"`
+	// Filters the resource list by the specified SupportStatus.
+	SupportStatus SupportStatus `q:"support_status"`
+	// Includes a description of the resource.
+	IncludeDescription bool `q:"with_description"`
+}
+
+// ToResourceTypeListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToResourceTypeListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	return q.String(), err
+}
+
+// ReturnDescription returns whether the caller is expecting resource type
+// descriptions to be returned from the API.
+func (opts ListOpts) ReturnDescription() bool {
+	return opts.IncludeDescription
+}
+
+// List makes a request against the API to list available resource types.
+func List(client *gophercloud.ServiceClient, opts ListOptsBuilder) (r ListResult) {
+	url := listURL(client)
+	if opts != nil {
+		r.hasDescriptions = opts.ReturnDescription()
+		query, err := opts.ToResourceTypeListQuery()
+		if err != nil {
+			r.Err = err
+			return
+		}
+		url += query
+	}
+	_, r.Err = client.Get(url, &r.Body, nil)
+	return
+}
+
+// Schema retreives the schema for a given resource type.
+func Schema(client *gophercloud.ServiceClient, resourceType string) (r SchemaResult) {
+	_, r.Err = client.Get(schemaURL(client, resourceType), &r.Body, nil)
+	return
+}
+
+// TemplateOptsBuilder allows extensions to add additional parameters to the
+// Template request.
+type TemplateOptsBuilder interface {
+	ToTemplateQuery() (string, error)
+}
+
+// TemplateOpts allows the filtering of collections through the API.
+type TemplateOpts struct {
+	UseCFn bool
```

Comment:
In 2019 there's literally no reason to ever generate a CloudFormation-style template, so I'd prefer to leave it as a boolean that defaults to HOT so that it can just be ignored.

Reply:
I completely understand this, too. Similar to my other comment, though, we want Gophercloud to model the API as closely as possible. You're more than welcome to leverage gophercloud/utils to implement a set of updated/modern/sane Heat functions.